shared-bindings/pixelbuf/PixelBuf.c

```diff
@@ -0,0 +1,506 @@
+                    items[0] = MP_OBJ_NEW_SMALL_INT(src_items[i]);
+                    items[1] = MP_OBJ_NEW_SMALL_INT(src_items[i+1]);
+                    items[2] = MP_OBJ_NEW_SMALL_INT(src_items[i+2]);
+                    if (self->bpp == 4)
+                        items[3] = MP_OBJ_NEW_SMALL_INT(src_items[i+3]);
+
+                    mp_obj_t *tuple = mp_obj_new_tuple(self->bpp, items);
+                    pixelbuf_set_pixel(destbuf + i, tuple, self->byteorder, self->bpp);
+                    // this should probbly be optimized and refactored
+                    if (self->two_buffers) {
+                        for (uint j = 0; j < self->bpp; j++) {
+                            adjustedbuf[i + j] = (destbuf[i + j] * self->brightness);
+                        }
+                    } else {
+                        for (uint j = 0; j < self->bpp; j++) {
+                            destbuf[i + j] = (destbuf[i + j] * self->brightness);
+                        }
+                    }
+                }
+                return mp_const_none;
+                #else
+                return MP_OBJ_NULL; // op not supported
+                #endif
+            } else {
+                // Read slice.
+                size_t len = slice.stop - slice.start;
+                mp_get_buffer_raise(self->two_buffers ? self->rawbytearray : self->bytearray, &bufinfo, MP_BUFFER_READ);
+                //return mp_obj_new_bytearray(len, (uint8_t *) bufinfo.buf + slice.start);
+                return pixelbuf_get_pixel_array((uint8_t *) bufinfo.buf + slice.start, len, self->byteorder, self->bpp);
+            }
+#endif
+        } else {
+            // Single index rather than slice.
+            size_t index = mp_get_index(self->base.type, self->pixels, index_in, false);
+            size_t offset = (index * self->bpp);
+            if (offset + self->bpp > self->bytes) 
+                mp_raise_IndexError("Pixel beyond bounds of buffer");
+            uint8_t *pixelstart = (uint8_t *)(self->two_buffers ? self->rawbuf : self->buf) + offset;
+            if (value == MP_OBJ_SENTINEL) {
+                // load
+                return pixelbuf_get_pixel(pixelstart, self->byteorder, self->bpp);
+            } else {
+                // store
+                pixelbuf_set_pixel(pixelstart, value, self->byteorder, self->bpp);
+                if (self->two_buffers) {
+                    uint8_t *adjustedstart = self->buf + offset;
+                    for (uint j = 0; j < self->bpp; j++) {
+                        adjustedstart[j] = (pixelstart[j] * self->brightness);
+                    }
+                } else {
+                    for (uint j = 0; j < self->bpp; j++) {
+                        pixelstart[j] = (pixelstart[j] * self->brightness);
+                    }
+                }
+                return mp_const_none;
+            }
+        }
+    }
+}
+
+void pixelbuf_set_pixel_int(uint8_t *buf, mp_int_t value, uint byteorder, uint bpp) {
```

Comment:
These five functions should be in shared-module instead. This file should only contain the conversion from Python to C.